index.rst

```diff
@@ -1,51 +1,1166 @@
+
+            self.assertRaises(pexExcept.OutOfRangeError,
+                              lambda: afwMath.makeInterpolate(np.array([0], dtype=float), np.array([1], dtype=float),
+                                                              afwMath.Interpolate.LINEAR))
+
+
+    class TestMemory(lsst.utils.tests.MemoryTestCase):
+        pass
+
+    def setup_module(module):
+        lsst.utils.tests.init()
+
+    if __name__ == "__main__":
+        lsst.utils.tests.init()
+        unittest.main()
+
+Here we see that there is only one class called from this test: ``lsst::afw::math::Interpolate``. We make sure to add the appropriate lines to ``mathLib.py``, ``Sconscript``, and ``interpolate.cc`` as we saw in :ref:`new_cpp`.
+
+Smart Pointers
+^^^^^^^^^^^^^^
+
+We declare the class in the standard way, adding
+
+.. code-block:: c++
+
+    py::class_<Interpolate, std::shared_ptr<Interpolate>> clsInterpolate(mod, "Interpolate");
+
+to the module section of ``interpolate.cc``.
+``Interpolate`` itself is a virtual class and makes use of a shared pointer (see :ref:`virtual_functions`), so notice that here we added ``std::shared_ptr<Interpolate>`` as an inherited class, which is necessary to access Interpolate as a pointer.
+
+Enum types
+^^^^^^^^^^
+
+The first method is an enum called ``Style``.
+We declare a value for each keyword that points to the corresponding value in the header file, with an ``export_values()`` method at the end:
+
+.. code-block:: c++
+
+    py::enum_<Interpolate::Style>(clsInterpolate, "Style")
+        .value("UNKNOWN", Interpolate::Style::UNKNOWN)
+        .value("CONSTANT", Interpolate::Style::CONSTANT)
+        .value("LINEAR", Interpolate::Style::LINEAR)
+        .value("NATURAL_SPLINE", Interpolate::Style::NATURAL_SPLINE)
+        .value("CUBIC_SPLINE", Interpolate::Style::CUBIC_SPLINE)
+        .value("CUBIC_SPLINE_PERIODIC", Interpolate::Style::CUBIC_SPLINE_PERIODIC)
+        .value("AKIMA_SPLINE", Interpolate::Style::AKIMA_SPLINE)
+        .value("AKIMA_SPLINE_PERIODIC", Interpolate::Style::AKIMA_SPLINE_PERIODIC)
+        .value("NUM_STYLES", Interpolate::Style::NUM_STYLES)
+        .export_values();
+
+.. warning::
+
+    Do not forget to add the ``.export_values()`` at the end or your enumerated types will not be added to the class!
+
+.. _virtual_functions:
+
+Virtual Functions and abstract Classes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Notice that ``Interpolate`` is an abstract class that cannot be called directly.
+Through examination of ``testInterpolate.py`` we see that ``Interpolate`` objects are created by using the ``makeInterpolate`` function, which is of type ``PTR(Interpolate)``.
+We will wrap ``makeInterpolate`` in :ref:`function_kwargs` but first we finish wrapping ``Interpolate``.
+The main function is the method ``interpolate``, which can be called with a double, list, or ndarray, however calling interpolate with a double is actually a virtual function, so we cannot wrap it directly.
+
```

Comment:
Talking about the overloads is potentially confusing, since the use of a lambda is because the function is pure virtual, which has nothing to do with overloading. Maybe reword as something like "we need to wrap `interpolate(double)`"?